Adds feature to merge Instrument Track patterns

IanCaio opened the pull request:
This PR implements a new action on the context menu of TrackContentObjectViews for merging note patterns. It's only applicable for InstrumentTrack's TCOs. In terms of changes on the code those were the main ones:

- Create a `getPattern` method on the `PatternView` so we can add the notes to a pattern obtained from a pattern view.
- Adds a context menu action called Merge and the appropriate context menu entry if the selection can be merged.
- Create two methods on TrackContentObjectView: `canMergeSelection` and `mergeTCOs`. The first check if a vector of TCOVs can be merged (the conditions: more than one TCO being selected, they all belonging to an Instrument Track and they all belonging to the same track). The second does the actual merging, by creating a new TCO and adding all the notes to it while removing the merged TCOs.
- Small note: There's a `TODO:` comment regarding a statement that should be removed once #5699 is merged.

There are a few warnings that might be unnecessary (they cover some situations that should not normally happen). A placeholder icon was created for the action, which could be later substituted by a better one.

![Screenshot from 2020-10-04 20-59-16](https://user-images.githubusercontent.com/22241596/95030565-181f7b00-0687-11eb-9a1e-6605bb55ad28.png)
![Screenshot from 2020-10-04 20-58-09](https://user-images.githubusercontent.com/22241596/95030574-1d7cc580-0687-11eb-9a32-ee28afd6675d.png)

IanCaio committed:
Initial Commit

	This branch aims to create a feature for merging Instrument Track's patterns in a single pattern.

	Current changes:
		- Adds a getPattern method to PatternView.
		- Adds a context menu action called "Merge".
		- Inside TrackContentObjectView::contextMenuEvent, we check if the selected TCOs can be merged and store the answer in a boolean called canMergeTCOs. The conditions to merge TCOs are: Having multiple TCOs selected; All TCOs belonging to a InstrumentTrack; All TCOs belonging to the same track.
		- If canMergeTCOs is true, add an action on the context menu for merging the selection.
		- Creates a method called TrackContentObjectView::mergeTCOs( tcovs ). That method first finds the earliest MidiTime from all selected TCOs and creates a pattern on that position. Then, notes are added to that pattern and repositioned depending on the relative position from the current TCOV being merged in relation to the new pattern. Some warnings were added to this method in case dynamic_casts failed, though they shouldn't. After all notes being added and positioned, newPattern->updateLength() and newPattern->dataChanged() are called.

	Known Bugs:
		- Some added notes don't play until you make a change (i.e.: move them) in the PianoRoll.

	To Do:
		- Remove QDebug include on Track.cpp.
		- Remove some debugging warnings.
		- Change the PixMap for the Merge Selection action on the context menu.
		- Fix bugs.

IanCaio committed:
Merge branch 'master' into feature/mergeTCOs

IanCaio committed:
Fix bug about notes not being played

	This commit fixes a previous bug where some notes wouldn't play after a merge operation happened. Apparently the root cause was Pattern->rearrangeAllNotes() not being called. It seems to be crucial that the notes are arranged in the right order for them to play properly.
	Removes QDebug header and includes Note.h.

IanCaio committed:
Reorganizes code to make it ready for review

	This commit reorganizes the code a little bit to make it better suited for review. The code that checked whether a selection can be merged was extracted to a separate method called TrackContentObjectView::canMergeSelection(). Two conditionals inside that code were merged into a single one, since they both resulted in the same action.
	Also updated a comment next to a line of code that is probably going to be removed after the PR #5699 is merged (it's currently necessary because of a bug that is fixed on that PR).

IanCaio committed:
Updates the icon for the merge action

superpaik commented:
[Note: this is my first review, so let me know if you need more detail]

The averall functionality works fine with me. Some issues, though:
- when you merge two patterns and one is empty, if it's the first the empty one, it merge the two of them, but if the empty is the second, it doesn't merge. I don't know if this behaviour is wanted or not, but seems odd to me.
- the functionality of undo (ctrl-z) is weird. If you merge two patterns, you need two undos to undo the merge. If you merge the patterns again (after the undo), then you need three undos. If you do it again, then four, and so on. If you merge more than two, say three, and the pattern to be merge are created in order (e.g. first on beat 5, then on beat 9, another on beat 13) and merge them, it works well (aside from the double undo), but if you create them and then move the pattern from 13 to 1 and merge and undo, the behaviour is weird (first the notes on the pattern on beat1 desappear, and then the pattern) ending with only patterns on beats 5 and 9.

IanCaio commented:
Thanks for testing and giving feedback!

> * when you merge two patterns and one is empty, if it's the first the empty one, it merge the two of them, but if the empty is the second, it doesn't merge. I don't know if this behaviour is wanted or not, but seems odd to me.

That is expected because the clip's size is recalculated so it is the shorted possible on the song editor (that also happens on the piano roll, if you delete notes from the last bar the clip will be shortened to have the fewest bars it can). So when you merge an empty clip with a non-empty one on that order it doesn't get shortened, because there are notes at the end of the clip. But if you merge a non-empty clip with an empty one the empty bars at the end are ignored and the clip is shortened to be the smallest size possible. They are still being merged though.

> * the functionality of undo (ctrl-z) is weird. If you merge two patterns, you need two undos to undo the merge. If you merge the patterns again (after the undo), then you need three undos. If you do it again, then four, and so on. If you merge more than two, say three, and the pattern to be merge are created in order (e.g. first on beat 5, then on beat 9, another on beat 13) and merge them, it works well (aside from the double undo), but if you create them and then move the pattern from 13 to 1 and merge and undo, the behaviour is weird (first the notes on the pattern on beat1 desappear, and then the pattern) ending with only patterns on beats 5 and 9.

Oh yeah, I actually didn't write a history system for merging, so the undo behavior you're seeing is being caused by some calls that are made inside the merge operation (`createTCO()`, `tcov->remove()` and I guess `addNotes()`). Merging works this way:

1) A new empty clip is created
2) There's a loop going through each merged clip, writing their notes on the new clip and deleting them

When you merge 2 clips and hit CTRL+Z, it first undo the clips being removed and then the new empty clip being created. You can confirm it by hitting CTRL+Z only once and then moving the clip: you'll see the two clips you had before below it.

I'm not sure it's worth it trying to fix it, specially because the history system is currently in a very unstable state already (I'd advise against using it at all and instead constantly saving the project as a new version) and is among the plans of things to be changed in the future. But I'll take a better look at it!

Spekular commented:
You should get most of the way to a correct undo by calling `addChecpoint` and `saveJournallingState(false)` at the start of the merge function, then `restoreJournallingState` at the end. I think it'll be a while until the undo/redo system is overhauled, and until it is we can't just let undo support devolve into a useless state.

IanCaio committed:
Merge branch 'master' into feature/mergeTCOs

IanCaio commented:
You're right, I'll try to make it work here!

IanCaio committed:
Fixes history system for the merge operation

	Fixes the history system for the merge operation by adding a journal checkpoint on the track where the merge is happening and disabling the journalling state of everything being handled during the operation, restoring the journalling state at the end.

IanCaio commented:
The history system was fixed for the merge operation, undo and redo should work fine now. I've a small doubt whether it's a good idea to restore the journalling state of a TCO after removing it, when I disabled it before removing. For now, I'm restoring the state but I left a TODO comment to remind me of removing the method call if it's unnecessary.

superpaik commented:
It works fine now, both undo as redo. I've tried with like 100 different patterns and it worked!

IanCaio commented:
Awesome, glad it's working now! Thanks for testing man!

IanCaio committed:
Merge branch 'master' into feature/mergeTCOs

	Fixes merge conflict on include/Track.h

IanCaio committed:
Merge branch 'master' into feature/mergeTCOs

IanCaio committed:
Removes method call not needed anymore

	After the bug fix from #5699, there's no need to call
movePosition on a TCO that was created with a position on its
constructor. The line that called movePosition was removed.
	Code style was also fixed.

IanCaio commented:
Merged master and removed line that is now unnecessary since the bug fix from #5699 .
Also fixed the code style.

IanCaio committed:
Merge branch 'master' into feature/mergeTCOs

IanCaio committed:
Updates PR to use TimePos

	After recently merge that renames MidiTime to TimePos, the code
was updated to use the new class name.

IanCaio committed:
Merge branch 'master' into feature/mergeTCOs

	Fixes conflict caused by #5806, which splitted Track.cpp and
Track.h

IanCaio committed:
Fixes pointer declaration code style

IanCaio committed:
Remove unnecessary call to restoreJournallingState

	After removing a TCO I was calling restoreJournallingState on
it, thinking it could be necessary for the journalling state be restored
when the history Undo brings the TCO back.
	It doesn't make much sense that this would be necessary, so I'm
removing it. After testing, it seems like restoring the TCO through Undo
brings it back with the journalling state restored.

IanCaio commented:
Fixed another conflict introduced by #5806 , changed the pointer declarations so they adhere to `Type* var;` code style and removed a line that I believe was unnecessary.

PhysSong commented on src/gui/TrackContentObjectView.cpp:
```diff
@@ -948,6 +955,8 @@ void TrackContentObjectView::contextMenuEvent( QContextMenuEvent * cme )
 
 	QMenu contextMenu( this );
 
+	bool canMergeTCOs = individualTCO ? false : canMergeSelection(selectedTCOs);
```
```suggestion
	bool canMergeTCOs = !individualTCO && canMergeSelection(selectedTCOs);
```

PhysSong commented on src/gui/TrackContentObjectView.cpp:
```diff
@@ -1106,6 +1127,122 @@ void TrackContentObjectView::toggleMute( QVector<TrackContentObjectView *> tcovs
+
+		// If there are TCOs from different tracks or TCOs from tracks
+		// other than an InstrumentTrack, can't merge them
+		if (ownerTrackView != previousOwnerTrackView
+			|| !dynamic_cast<InstrumentTrackView*>(ownerTrackView))
```
Instead of trying `dynamic_cast` for every `TrackView`s, you can check the type of the first `ownerTrackView` and then compare with the other `ownerTrackView`s without type checking.

IanCaio replied:
True! I'll move the `dynamic_cast` to the `if (!previousOwnerTrackView)` conditional and return false if it fails

Spekular commented on src/gui/TrackContentObjectView.cpp:
```diff
@@ -1106,6 +1127,122 @@ void TrackContentObjectView::toggleMute( QVector<TrackContentObjectView *> tcovs
 	}
 }
 
+bool TrackContentObjectView::canMergeSelection(QVector<TrackContentObjectView*> tcovs)
+{
+	// We can only merge InstrumentTrack's TCOs, so check if we only have those in the selection,
+	// and also if they all belong to the same track
+
+	// We start assuming we can merge and if any condition is broken we set it to false
+	bool canMerge = true;
+
+	// Variable to check if all TCOs belong to the same track
+	TrackView* previousOwnerTrackView = nullptr;
+
+	// Then we check every selected TCO to see if all of them are InstrumentTrack's TCOs.
+	// If any isn't, we set canMerge to false and quit the loop.
+	for (auto tcov: tcovs)
+	{
+		TrackView* ownerTrackView = tcov->getTrackView();
+
+		// Set the previousOwnerTrackView to the first TrackView
+		if (!previousOwnerTrackView)
+		{
+			previousOwnerTrackView = ownerTrackView;
+		}
+
+		// If there are TCOs from different tracks or TCOs from tracks
+		// other than an InstrumentTrack, can't merge them
+		if (ownerTrackView != previousOwnerTrackView
+			|| !dynamic_cast<InstrumentTrackView*>(ownerTrackView))
+		{
+			canMerge = false;
+			break;
+		}
+	}
+
+	return canMerge;
+}
```
Simplify the logic by outsourcing it to a set data structure? Probably no noticeable performance impact.
```suggestion
bool TrackContentObjectView::canMergeSelection(QVector<TrackContentObjectView*> tcovs)
{
	bool canMerge = tcovs.size() > 1 &&
		!dynamic_cast<InstrumentTrackView*>(tcovs.at(0)->getTrackView());

	QSet<TrackView*> parentTracks;
	for (auto tcov: tcovs){ parentTracks.insert(tcov); }

	return canMerge && parentTracks.size() == 1;
}
```

IanCaio replied:
Done! Used `std::set` instead and removed the `tcovs.size() > 1` since that was checked for on the `contextMenuAction`. Thinking about moving it here though, probably belongs better to this method.

Spekular replied:
Yeah, that's why I included it, feels incorrect for canMerge to return true for a single clip.

IanCaio replied:
Agreed, it wasn't a problem before because canMergeSelection was only called with multiple TCOs, but this check suits the method better. Changed on the last commit!

Spekular commented on src/gui/TrackContentObjectView.cpp:
```diff
@@ -1106,6 +1127,122 @@ void TrackContentObjectView::toggleMute( QVector<TrackContentObjectView *> tcovs
+	track->saveJournallingState(false);
+
+	// Find the earliest position of all the selected TCOVs
+	TimePos earliestPos = tcovs.at(0)->getTrackContentObject()->startPosition();
+	TimePos currentPos = earliestPos;
+
+	for (auto tcov: tcovs)
+	{
+		currentPos = tcov->getTrackContentObject()->startPosition();
+		if (currentPos < earliestPos)
+		{
+			earliestPos = currentPos;
+		}
+	}
```
Something like this should work?
```suggestion
	const auto earliestTCOV = std::min_element(tcovs.begin(), tcovs.end(),
		[](TrackContentObjectView* a, TrackContentObjectView* b){
			return a->getTrackContentObject()->startPosition() <
			b->getTrackContentObject()->startPosition;
		}
	);
	const TimePos earliestPos = earliestTCOV->getTrackContentObject()->startPosition();
```

IanCaio replied:
Are Qt iterators guaranteed to be compatible with methods that expect `std::iterator`s?

Spekular replied:
I think so? https://doc.qt.io/qt-5/containers.html#stl-style-iterators

These can probably be constBegin and constEnd by the way.

IanCaio replied:
Cool! I thought they should be, but was asking to make sure. I'm trying to use `std::set` on the other suggestion btw, just to follow that guideline of avoiding Qt where it isn't needed (even though `QVector` is being used for the selection, but that's something for another PR :laughing: )

Spekular reviewed (commented):
LGTM. Suggested a few potential changes to make things shorter, but I haven't tested them to ensure they're totally correct so there might be syntax issues and such.

IanCaio committed:
Apply PhysSong suggestions

	Change canMergeTCOs to use logical operators instead of the
ternary operator.
	Avoids repeating the dynamic_cast unnecessarily on the
canMergeSelection method.

IanCaio committed:
Applies Spekular's suggestions

	Saves lots of lines on canMergeSelection by using a data set
to store all the TCO owner tracks, so we can later check how many tracks
we have.
	Uses std::min_element to find the earliest TCOV and get it's
starting position.

IanCaio committed:
canMergeSelection now checks number of TCOs

	The logic that returned we couldn't merge a single TCO was
inside contextMenuEvent, when it suited more the canMergeSelection
method. That was moved there and now even if canMergeSelection receives
a list with just a single TCO it will return that it can't be merged.

PhysSong commented on src/gui/TrackContentObjectView.cpp:
```diff
@@ -1106,6 +1127,100 @@ void TrackContentObjectView::toggleMute( QVector<TrackContentObjectView *> tcovs
+	// Restore journalling states now that the operation is finished
+	track->restoreJournallingState();
+	newPattern->restoreJournallingState();
+	// Update engine
```
```suggestion
	// Update song
```

PhysSong commented on src/gui/TrackContentObjectView.cpp:
```diff
@@ -1106,6 +1127,100 @@ void TrackContentObjectView::toggleMute( QVector<TrackContentObjectView *> tcovs
+	// Rearrange notes because we might have moved them
+	newPattern->rearrangeAllNotes();
+	// Restore journalling states now that the operation is finished
+	track->restoreJournallingState();
+	newPattern->restoreJournallingState();
```
Not important, but I want to change the order of two calls because `saveJournallingState`/`restoreJournallingState` works like push/pop.
```suggestion
	newPattern->restoreJournallingState();
	track->restoreJournallingState();
```

IanCaio replied:
Done! Does that mean that when I called `track->restoreJournallingState();` I was actually restoring `newPattern`s journalling state?

PhysSong replied:
It's rather a stylistic/semantic point, nothing was working wrong.

IanCaio replied:
Oh, got it now! Thanks!

IanCaio committed:
Changes order of restoreJournallingState calls

	Calls restoreJournallingState respecting the push/pop order in
relation to the saveJournallingState calls.

	Fix a comment.

IanCaio commented:
Sure, I think it's good to go! And it has 2 approvals now :smile:
